### PR TITLE
feat: add expression rule support

### PR DESCRIPTION
**Pull request by StellarisW**
#### What type of PR is this?

feat

#### What this PR does / why we need it (en: English/zh: Chinese):

en: add permission checks mode - "CUSTOM"
zh:  添加 permission 检测模式 - "自定义"，用于解析 "A && B || C", "A && !B"等场景

### CODE REVIEW

File: option.go [hunk 1]
```diff
@@ -91,10 +98,11 @@ var OptionsDefault = Options{
 
 func NewOptions(opts ...Option) *Options {
 	options := &Options{
-		Logic:            OptionsDefault.Logic,
-		PermissionParser: OptionsDefault.PermissionParser,
-		Unauthorized:     OptionsDefault.Unauthorized,
-		Forbidden:        OptionsDefault.Forbidden,
+		Logic:               OptionsDefault.Logic,
+		PermissionParser:    OptionsDefault.PermissionParser,
+		PermissionSeparator: DefaultPermissionSeparator,
```
Comment (L2ncE) [inline option.go]: Using `OptionsDefault.PermissionSeparator`

File: option.go [hunk 2]
```diff
@@ -56,7 +57,8 @@ type Options struct {
 	// PermissionParserFunc is used for parsing the permission
 	// to extract object and action usually
 	// Optional. Default: PermissionParserWithSeparator(":")
-	PermissionParser PermissionParserFunc
+	PermissionParser    PermissionParserFunc
+	PermissionSeparator string
```
Comment (L2ncE) [inline option.go:62]: Add comment

File: casbin.go [hunk 3]
```diff
@@ -16,6 +16,8 @@ package casbin
 
 import (
 	"context"
+	"github.com/Knetic/govaluate"
```
Comment (L2ncE) [inline casbin.go]: Use two-stage & Can we not import third-party libraries

File: casbin.go [hunk 4]
```diff
@@ -111,18 +115,66 @@ func (m *Middleware) RequiresPermissions(permissions []string, opts ...Option) a
...
+			if res, ok := result.(bool); !ok {
+				c.AbortWithStatus(consts.StatusInternalServerError)
+				return
+			} else {
+				if !res {
+					options.Forbidden(ctx, c)
+					return
+				}
+			}
+
```
Comment (L2ncE) [inline casbin.go:164]: Delete useless blank

File: casbin.go [hunk 5]
```diff
@@ -158,6 +212,42 @@ func (m *Middleware) RequiresRoles(requiredRoles []string, opts ...Option) app.H
...
```
Comment (L2ncE) [inline casbin.go:249]: ditto

**Comment by L2ncE** [comment]
Can you restore the code in the original example and create a new folder in the example to put the `CUSTOM` sample?REAMDE is the same.

File: README.md [hunk 6]
```diff
@@ -81,10 +94,34 @@ func subjectFromSession(ctx context.Context, c *app.RequestContext) string {
...
+| ------------------------- | ------------------------------------------------------------ | ------------------------------------------------------------ |
+| Logic                     | `AND`                                                        | Logic is the logical operation (AND/OR/CUSTOM) used in permission checks in case multiple permissions or roles are specified. |
+| PermissionParser          | `PermissionParserWithSeparator(":")`                         | PermissionParserFunc is used for parsing the permission to extract object and action usually. |
+| PermissionParserSeparator | `*`                                                          | PermissionParserSeparator is used for parsing the permission to extract object and action usually. |
+| Unauthorized              | `func(ctx context.Context, c *app.RequestContext) {    c.AbortWithStatus(consts.StatusUnauthorized) }` | Unauthorized defines the response body for unauthorized responses. |
+| Forbidden                 | `func(ctx context.Context, c *app.RequestContext) {    c.AbortWithStatus(consts.StatusForbidden) }` | Forbidden defines the response body for forbidden responses. |
+
+- `WithLogic(logic Logic)`
```
Comment (L2ncE) [inline README.md]: You can put these in the documentation and leave only the critical parts here

File: casbin.go [hunk 7]
```diff
@@ -111,18 +115,66 @@ func (m *Middleware) RequiresPermissions(permissions []string, opts ...Option) a
...
+			expression = strings.Replace(expression, options.PermissionSeparator, "\\"+options.PermissionSeparator, -1)
+			exp, err := govaluate.NewEvaluableExpression(expression)
+			if err != nil {
+				c.AbortWithStatus(consts.StatusInternalServerError)
+				return
+			}
+
+			permissions := exp.Vars()
+			params := make(map[string]interface{}, len(permissions))
+
```
Comment (L2ncE) [inline casbin.go:129]: Use `utils.H` instead of `map[string]interface{}`

**Review by L2ncE — CHANGES_REQUESTED** [review]

File: casbin.go [hunk 8]
```diff
@@ -16,7 +16,10 @@ package casbin
 
 import (
 	"context"
+	"github.com/cloudwego/hertz/pkg/common/utils"
```
Comment (L2ncE) [inline casbin.go]: two-stage

**Review by L2ncE — APPROVED** [review]

**Comment by li-jin-gou** [comment]
LGTM

**Pull request merged**